guides/internals.rst

```diff
@@ -0,0 +1,342 @@
+normal order on strings - find the first byte at which they differ and use that
+to decide) smallest.
+
+Ideally we could think of the shrinker as a generic function that takes a
+string satisfying some predicate and returns the shortlex minimal string that
+also satisfies it.
+This is wrong on several levels: The first is that we only succeed in approximating
+such a minimal string. The second is that we are only interested in minimizing
+things where the predicate goes through the Hypothesis API, which lets us track
+a lot of info about how the data is used and use that to guide the process.
+
+We then use a number of different transformations of the string to try and
+reduce our input. These vary from principled general transformations to shameless
+hacks that special case something we need to work well. We try to aim for mostly
+the former, but the nice thing about this model is that the underlying representation
+is fully general and we are free to try whatever we want and it will never result
+in us doing the wrong thing, so hacks are only a problem to the degree that they
+result in messy code and fragile heuristics, they're never a correctness issue,
+so if we can't make something work without such a hack it's not a big deal.
+
+One such example of a hack is the handling of floating point numbers. There are
+a couple of lexicographic shrinks that are always valid but only really make
+sense for our particular encoding of floats. We check if we're working
+on something that is of the right size to be a float and apply those
+transformations regardless of whether it is actually meant to be a float.
+Worst case scenario it's not a float and they don't work, and we've run a few
+extra test cases.
+
+--------------------------
+Useful Files to Know About
+--------------------------
+
+The code associated with Conjecture lives in
+`src/hypothesis/internal/conjecture <https://github.com/HypothesisWorks/hypothesis-python/tree/master/src/hypothesis/internal/conjecture>`_.
+There are a number of files in there,
+but the most important ones are ``engine.py`` and ``data.py``.
+``data.py`` defines the core type that is used to represent test cases,
+and ``engine.py`` contains the main driver for deciding what test cases to run.
+
+There is also ``minimizer.py``, which contains a general purpose lexicographic
+minimizer. This is responsible for taking some byte string and a predicate over
+byte strings and producing a string of the same length which is lexicographically
+smaller. Unlike the shrinker in general, this *is* supposed to work on arbitrary
+predicates and doesn't know anything about the testing API. We typically apply
+this to subsets of the bytes for a test input with a predicate that knows how
+to integrate those subsets into a larger test. This is the part of the code
+that means we can do things like replacing an integer with a smaller one.
+
+-------
+Testing
```

Comment:
This should probably be moved to `testing-hypothesis.rst`, and just leave a reference here.

We should also explain how to install the test dependencies.